src/app/components/user-form/user-form/user-form.component.html

```diff
@@ -0,0 +1,65 @@
+<ng-container [formGroup]="userFormGroup">
+    <div>
+        <div>
+            <label for="firstName">First Name</label>
+            <input type="text" name="firstName" formControlName="firstName">
+            <span *ngIf="userFormGroup.get('firstName')?.errors?.['required']" class="error-message">Field is
+                required</span>
+        </div>
+        <div>
+            <label for="lastName">Last Name</label>
+            <input type="text" name="lastName" formControlName="lastName">
+            <span *ngIf="userFormGroup.get('lastName')?.errors?.['required']" class="error-message">Field is
+                required</span>
+        </div>
+        <div>
+            <label for="age">Age</label>
+            <input type="number" name="age" formControlName="age">
+            <span *ngIf="userFormGroup.get('age')?.errors?.['required']" class="error-message">Field is required</span>
```

Comment:
It is better to use userFormGroup.controls.age instead of userForm.get('age').

Accessing via controls makes type checking.

It helps to avoid optional chain (userFormGroup.get('age')?.hasError('required'), you can use userForm.controls.age.hasError('required')

It also makes to use hasError() instead of errors?.[key]

You can also create getter for "age" form control to simplify usage in template:

`get ageControl(): FormControl<number> {
  return this.userFormGroup.controls.age;
}`

or use @let in template:

https://angular.dev/api/core/@let